lib/src/zenohUri.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ * Copyright (c) 2024 General Motors GTO LLC
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * SPDX-FileType: SOURCE
+ * SPDX-FileCopyrightText: 2024 General Motors GTO LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+ 
+#include <iomanip>
+#include <sstream>
+#include <spdlog/spdlog.h>
+#include <up-cpp/uri/serializer/MicroUriSerializer.h>
+#include "up-client-zenoh-cpp/session/zenohSessionManager.h"
+
+namespace uprotocol::uri {
+
+namespace {
+void hexlify(auto& topic, const uint8_t byte) {
+    topic << std::hex << std::setw(2) << std::setfill('0');
+    topic << static_cast<int>(byte);
+}
+
+void hexlify(auto& topic, const auto &&start, const auto &&end) {
+    for (auto i = start; i < end; ++i) {
+        hexlify(topic, *i);
+    }
+}
+
+} // anonymous namespace
+
+std::string toZenohKeyString(const uprotocol::v1::UUri &u_uri) {
```

Comment:
Looking into it. The coverage space looks fairly small - should there also be tests for IPv6 and IPv4 authorities?

Reply:
Probably! If you add them here we can ask @evshary to do the same in `up-client-zenoh-rust` :slightly_smiling_face:

Reply:
Tests have been added, including several new tests.

Reply:
Thanks @gregmedd -- I added an issue on `up-client-zenoh-rust`:
https://github.com/eclipse-uprotocol/up-client-zenoh-rust/issues/23
in order to align.

The idea being we should probably have the same tests and expected results for both clients given they need to ineroperate.

Reply:
@gregmedd Yes, I agree with you. I'll also add tests for both IPv4 and IPv6.
BTW, although it might need some time to be merged, we plan to update the way to transform UUri to Zenoh key
https://github.com/eclipse-uprotocol/up-spec/pull/103

There are two aspects:
1. Use short-form UUri instead of micro-form UUri
2. Do not differentiate between local and remote UUri. Zenoh Key will need to add UAuthority in both cases.

Reply:
@evshary Sounds good. I think that the test inputs (UUri objects) will probably still be valid after the switch to short-form serialization, with only the check values needing to be updated.